planner/core/optimizer.go

```diff
@@ -517,62 +517,118 @@ func physicalOptimize(logic LogicalPlan, planCounter *PlanCounterTp) (plan Physi
 
 // eliminateUnionScanAndLock set lock property for PointGet and BatchPointGet and eliminates UnionScan and Lock.
 func eliminateUnionScanAndLock(sctx sessionctx.Context, p PhysicalPlan) PhysicalPlan {
-	var pointGet *PointGetPlan
-	var batchPointGet *BatchPointGetPlan
-	var physLock *PhysicalLock
-	var unionScan *PhysicalUnionScan
-	iteratePhysicalPlan(p, func(p PhysicalPlan) bool {
-		if len(p.Children()) > 1 {
-			return false
-		}
-		switch x := p.(type) {
-		case *PointGetPlan:
-			pointGet = x
-		case *BatchPointGetPlan:
-			batchPointGet = x
-		case *PhysicalLock:
-			physLock = x
-		case *PhysicalUnionScan:
-			unionScan = x
-		}
-		return true
-	})
-	if pointGet == nil && batchPointGet == nil {
+	type Tuple struct {
+		// Point get and batch point get cannot be not nil simultaneously
+		pointGet      *PointGetPlan
+		batchPointGet *BatchPointGetPlan
+
+		physicalLock *PhysicalLock
+	}
+
+	tuples := make([]*Tuple, 0)
+
+	// physicalLockMap is a map for mapping PhysicalLock to PhysicalUnionScan. We need this map because when
+	// getLockWaitTime(sctx, physicalLock.Lock) returns false, we delete the relevant PhysicalUnionScan (if it exists)
+	// in the unionScanSet to avoid elimination.
+	physicalLockMap := make(map[PhysicalPlan]PhysicalPlan)
+	unionScanSet := make(map[PhysicalPlan]interface{})
+
+	iteratePhysicalPlan(p, nil, nil,
+		func(p PhysicalPlan, physicalLock *PhysicalLock, unionScan *PhysicalUnionScan) (*PhysicalLock, *PhysicalUnionScan, bool) {
+			switch x := p.(type) {
+			case *PointGetPlan:
+				tuples = append(tuples, &Tuple{
+					pointGet:     x,
+					physicalLock: physicalLock,
+				})
+				if physicalLock != nil {
+					physicalLockMap[physicalLock] = unionScan
+				}
+				if unionScan != nil {
+					unionScanSet[unionScan] = nil
+				}
+			case *BatchPointGetPlan:
+				tuples = append(tuples, &Tuple{
+					batchPointGet: x,
+					physicalLock:  physicalLock,
+				})
+				if physicalLock != nil {
+					physicalLockMap[physicalLock] = unionScan
+				}
+				if unionScan != nil {
+					unionScanSet[unionScan] = nil
+				}
+			case *PhysicalLock:
+				return x, nil, false
+			case *PhysicalUnionScan:
+				return nil, x, false
+			}
+			return nil, nil, false
+		})
+	if len(tuples) == 0 {
 		return p
 	}
-	if physLock == nil && unionScan == nil {
+	if len(physicalLockMap) == 0 && len(unionScanSet) == 0 {
 		return p
 	}
-	if physLock != nil {
-		lock, waitTime := getLockWaitTime(sctx, physLock.Lock)
-		if !lock {
-			return p
-		}
-		if pointGet != nil {
-			pointGet.Lock = lock
-			pointGet.LockWaitTime = waitTime
-		} else {
-			batchPointGet.Lock = lock
-			batchPointGet.LockWaitTime = waitTime
+
+	if len(physicalLockMap) != 0 {
+		for _, tuple := range tuples {
+			if tuple.physicalLock == nil {
+				continue
+			}
+
+			lock, waitTime := getLockWaitTime(sctx, tuple.physicalLock.Lock)
```

Comment:
I think the below plan is not right with the current implement. The table scan's lock was lost.
```
mysql> desc select * from (select * from t union all select * from t where id=1) t1 for update;
+----------------------------+---------+-----------+---------------+--------------------------------+
| id                         | estRows | task      | access object | operator info                  |
+----------------------------+---------+-----------+---------------+--------------------------------+
| Union_13                   | 3.00    | root      |               |                                |
| ├─TableReader_16           | 2.00    | root      |               | data:TableFullScan_15          |
| │ └─TableFullScan_15       | 2.00    | cop[tikv] | table:t       | keep order:false, stats:pseudo |
| └─Point_Get_18             | 1.00    | root      | table:t       | handle:1, lock                 |
+----------------------------+---------+-----------+---------------+--------------------------------+
4 rows in set (0.00 sec)
```